### PR TITLE
Bi-directional DEX fee burn capability

### LINKED ISSUE #1030: Add support for bi-directional and governance controlled DEX fee
TODO

### PR DESCRIPTION
**Pull request by bvbfan**
/kind feature

Fixes #1030

### CODE REVIEW

File: src/masternodes/govvariables/attributes.cpp [hunk 1]
```diff
@@ -134,39 +150,53 @@ UniValue ATTRIBUTES::Export() const {
...
         if (strVec.empty()) {
             return Res::Err("Empty map key found");
         }
 
-        // switch for the many more types coming later!
-        if (strVec[0].size() == 1 && strVec[0].at(0) == AttributeTypes::Token) {
-            if (strVec.size() != 3) {
-                return Res::Err("Three items expected in token key");
-            }
+        if (strVec[0].size() != 1) {
```
Comment (Bushstar) [inline src/masternodes/govvariables/attributes.cpp:167]: Does not have to be 1, for token it is set to 1 but for others a longer string could be used.

File: src/masternodes/govvariables/attributes.cpp [hunk 2]
```diff
@@ -182,5 +250,24 @@ Res ATTRIBUTES::Validate(const CCustomCSView & view) const
...
+        auto res = ProcessVariable(item.first, item.second, GetTypeByKey, GetTypeByKey,
+                                   [&](const UniValue& values) -> Res {
+            if (values[0].get_int() == AttributeTypes::Token) {
+                if (values[2].get_int() == TokenKeys::DexFeePCT) {
+                    auto res = VerifyPct(values[3].getValStr());
+                    if (!res) {
+                        return std::move(res);
+                    }
+                    auto tokenId = uint32_t(values[1].get_int());
+                    return mnview.SetDexfeePct({tokenId}, *res.val);
```
Comment (Bushstar) [inline src/masternodes/govvariables/attributes.cpp:263]: I do not think this is correct. ATTRBIUTES is the store of data which should not be then stored separately like in other Gov vars. We just need a better way to access the map.

Comment (bvbfan) [reply src/masternodes/govvariables/attributes.cpp:263]: We can parse and keep in `UniValue` i still prefer to keep some of them separately as `LP_SPLITS::Apply` or `ICX_TAKERFEE_PER_BTC::Apply`

Comment (bvbfan) [reply src/masternodes/govvariables/attributes.cpp:263]: Actually all variables apply records to the particular view.

**Review by Bushstar — APPROVED** [review]

File: src/masternodes/mn_checks.cpp [hunk 3]
```diff
@@ -3598,11 +3598,22 @@ Res CPoolSwap::ExecuteSwap(CCustomCSView& view, std::vector<DCT_ID> poolIDs, boo
...
         // Perform swap
-        poolResult = pool->Swap(swapAmount, poolPrice, [&] (const CTokenAmount &tokenAmount) {
+        poolResult = pool->Swap(swapAmount, dexfeeInPct, poolPrice, [&] (const CTokenAmount& dexfeeInAmount, const CTokenAmount& tokenAmount) {
             // Save swap amount for next loop
             swapAmountResult = tokenAmount;
 
+            CTokenAmount dexfeeOutAmount{tokenAmount.nTokenId, 0};
+
+            if (height >= Params().GetConsensus().FortCanningHillHeight) {
+                if (auto dexfeeOutPct = view.GetDexfeePct(currentID, tokenAmount.nTokenId)) {
```
Comment (prasannavl) [inline src/masternodes/mn_checks.cpp:3611]:
```suggestion
                if (auto dexfeeOutPct = view.GetDexFeePct(currentID, tokenAmount.nTokenId)) {
```

File: src/masternodes/poolpairs.cpp [hunk 4]
```diff
@@ -695,3 +705,22 @@ void CPoolPairView::ForEachPoolShare(std::function<bool (DCT_ID const &, CScript
         return callback(poolShareKey.poolID, poolShareKey.owner, height);
     }, startKey);
 }
+
+Res CPoolPairView::SetDexfeePct(DCT_ID poolId, DCT_ID tokenId, CAmount feePct) {
```
Comment (prasannavl) [inline src/masternodes/poolpairs.cpp:709]:
```suggestion
Res CPoolPairView::SetDexFeePct(DCT_ID poolId, DCT_ID tokenId, CAmount feePct) {
```

File: src/masternodes/poolpairs.cpp [hunk 5]
```diff
@@ -695,3 +705,22 @@ void CPoolPairView::ForEachPoolShare(std::function<bool (DCT_ID const &, CScript
...
+        return Res::Err("No such pool pair");
+    }
+    if (feePct < 0 || feePct > COIN) {
+        return Res::Err("Token dex fee should be in percentage");
+    }
+    WriteBy<ByTokenDexfeePct>(std::make_pair(poolId, tokenId), uint32_t(feePct));
+    return Res::Ok();
+}
+
+CAmount CPoolPairView::GetDexfeePct(DCT_ID poolId, DCT_ID tokenId) const {
```
Comment (prasannavl) [inline src/masternodes/poolpairs.cpp:720]:
```suggestion
CAmount CPoolPairView::GetDexFeePct(DCT_ID poolId, DCT_ID tokenId) const {
```

File: src/masternodes/poolpairs.h [hunk 6]
```diff
@@ -231,6 +231,9 @@ class CPoolPairView : public virtual CStorageView
     Res SetRewardLoanPct(DCT_ID const & poolId, uint32_t height, CAmount rewardLoanPct);
     bool HasPoolPair(DCT_ID const & poolId) const;
 
+    Res SetDexfeePct(DCT_ID poolId, DCT_ID tokenId, CAmount feePct);
```
Comment (prasannavl) [inline src/masternodes/poolpairs.h:234]:
```suggestion
    Res SetDexFeePct(DCT_ID poolId, DCT_ID tokenId, CAmount feePct);
```

File: src/masternodes/poolpairs.h [hunk 7]
```diff
@@ -231,6 +231,9 @@ class CPoolPairView : public virtual CStorageView
...
+    CAmount GetDexfeePct(DCT_ID poolId, DCT_ID tokenId) const;
```
Comment (prasannavl) [inline src/masternodes/poolpairs.h:235]:
```suggestion
    CAmount GetDexFeePct(DCT_ID poolId, DCT_ID tokenId) const;
```

File: src/masternodes/poolpairs.h [hunk 8]
```diff
@@ -248,6 +251,7 @@ class CPoolPairView : public virtual CStorageView
     struct ByDailyLoanReward{ static constexpr uint8_t prefix() { return 'q'; } };
     struct ByRewardLoanPct  { static constexpr uint8_t prefix() { return 'U'; } };
     struct ByPoolLoanReward { static constexpr uint8_t prefix() { return 'W'; } };
+    struct ByTokenDexfeePct { static constexpr uint8_t prefix() { return 'l'; } };
```
Comment (prasannavl) [inline src/masternodes/poolpairs.h:254]:
```suggestion
    struct ByTokenDexFeePct { static constexpr uint8_t prefix() { return 'l'; } };
```

File: src/masternodes/gv.cpp [hunk 9]
```diff
@@ -82,11 +82,9 @@ void CGovView::EraseStoredVariables(const uint32_t height)
     }
 }
 
-std::map<std::string, std::string> CGovView::GetAttributes() const {
+std::shared_ptr<ATTRIBUTES> CGovView::GetAttributes() const {
     if (const auto var = GetVariable("ATTRIBUTES")) {
-        if (const auto attrs = dynamic_cast<ATTRIBUTES*>(var.get())) {
-            return attrs->attributes;
-        }
+        return std::dynamic_pointer_cast<ATTRIBUTES>(var);
```
Comment (Bushstar) [inline src/masternodes/gv.cpp:87]: This function should return the following map for direct use externally.

`std::map<CAttributeType, CAttributeValue> attributes;`

**Review by Bushstar — APPROVED** [review]

File: src/masternodes/gv.cpp (hunk 9 above)
Comment (Bushstar) [reply src/masternodes/gv.cpp:87]: Includes for this create recursive problem.

**Pull request merged**